### PR TITLE
Use StyleLayer in Bucket

### PR DESCRIPTION
**Pull request by lucaswoj**
... and then refactor `StyleLayer` a little 

ref #1957

### CODE REVIEW

File: js/style/style_layer.js [hunk 1]
```diff
@@ -34,44 +34,56 @@ function StyleLayer(layer, refLayer) {
...
 
 StyleLayer.prototype = {
-    resolveLayout: function() {
-        if (!this.ref) {
-            this.layout = new LayoutProperties[this.type](this._layer.layout);
-        }
-    },
 
     setLayoutProperty: function(name, value) {
-        if (value == null) {
```
Comment (jfirebaugh) [inline js/style/style_layer.js]: Setting a layout or paint property to `null` should have the effect of resetting the effective value to the default. Is this behavior preserved?

**Comment by jfirebaugh** [comment]
The layer data that is passed to the worker comes from `StyleLayer#json`, which uses `StyleLayer#layout` and `StyleLayer#paint`, which are the currently calculated values, not unevaluated functions. This is not what we want -- the worker needs unevaluated layout functions so it can calculate values at multiple zoom levels.

**Comment by lucaswoj** [comment]
I should have been more clear: this is still under development.

> The layer data that is passed to the worker comes from `StyleLayer#json`, which uses `StyleLayer#layout` and `StyleLayer#paint`, which are the currently calculated values, not unevaluated functions. This is not what we want -- the worker needs unevaluated layout functions so it can calculate values at multiple zoom levels.

I'm working on this now :smile:

File: js/style/style_layer.js (hunk 1 above)
Comment (lucaswoj) [inline js/style/style_layer.js]: Weirdly, fixing this behavior caused some [unit tests to break](https://github.com/mapbox/mapbox-gl-js/pull/1977/files#diff-df9dac9f0f6edf6f9675a223d7687c17R88). I think think we should consider the tests wrong and the intended behavior (i.e. your proposed change) correct.

File: js/data/symbol_bucket.js [hunk 2]
```diff
@@ -26,6 +26,18 @@ function SymbolBucket(options) {
...
+    // to use a text-size value that is the same for all zoom levels.
+    // This calculates text-size at a high zoom level so that all tiles can
+    // use the same value when calculating anchor positions.
+    var zoomHistory = { lastIntegerZoom: Infinity, lastIntegerZoomTime: 0, lastZoom: 0 };
+
+    this.adjustedTextMaxSize = this.layer.getLayoutValue('text-size', 18, zoomHistory);
+    this.adjustedTextSize = this.layer.getLayoutValue('text-size', this.zoom + 1, zoomHistory);
+
+    this.adjustedIconMaxSize = this.layer.getLayoutValue('icon-size', 18, zoomHistory);
+    this.adjustedIconSize = this.layer.getLayoutValue('icon-size', this.zoom + 1, zoomHistory);
```
Comment (lucaswoj) [inline js/data/symbol_bucket.js]: this code replaces [this](https://github.com/mapbox/mapbox-gl-js/pull/1977/files#diff-c1d5cefb566acef357121481a5a8e27eL176) code

File: test/js/style/style.test.js [hunk 3]
```diff
@@ -870,12 +866,12 @@ test('Style#featuresAt', function(t) {
...
-            t.test('includes arbitrary keys', function(t) {
+            t.test('includes metadata', function(t) {
                 featuresInOrAt({}, function(err, results) {
                     t.error(err);
 
                     var layer = results[0].layer;
-                    t.equal(layer.something, 'else');
+                    t.equal(layer.metadata.something, 'else');
 
                     t.end();
```
Comment (lucaswoj) [inline test/js/style/style.test.js]: :warning: This is a breaking change; however it has been against our style spec for a while to put arbitrary keys in layers for some time.

File: test/js/style/style_layer.test.js [hunk 4]
```diff
@@ -93,10 +77,9 @@ test('StyleLayer#setPaintProperty', function(t) {
             }
         });
 
-        layer.resolvePaint({});
         layer.setPaintProperty('background-color', null);
 
-        t.deepEqual(layer.getPaintProperty('background-color'), [0, 0, 0, 1]);
+        t.equal(layer.getPaintProperty('background-color'), undefined);
```
Comment (lucaswoj) [inline test/js/style/style_layer.test.js]: :warning: This is a breaking change; however the old behavior was "wrong"

# Old Behavior

```js
var layer1 = new StyleLayer({ ..., paint: { } });
layer.getPaintProperty('background-color') // undefined

var layer2 = new StyleLayer({ ..., paint: { 'background-color': 'red' } });
layer.setPaintProperty('background-color', null)
layer.getPaintProperty('background-color') // [0, 0, 0, 1]
```

# New Behavior

```js
var layer1 = new StyleLayer({ ..., paint: { } });
layer.getPaintProperty('background-color') // undefined

var layer2 = new StyleLayer({ ..., paint: { 'background-color': 'red' } });
layer.setPaintProperty('background-color', null)
layer.getPaintProperty('background-color') // undefined
```

**Comment by lucaswoj** [comment]
Ok, phew, last of the `StyleLayer` _refactoring_.

:eyes: @jfirebaugh @ansis @mourner

**Comment by lucaswoj** [comment]
Perf isn't looking great. Investigating.

```
duration: 3000
43 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.12.4/mapbox-gl.js
37.3 fps /dist/mapbox-gl.js
45.5 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.12.4/mapbox-gl.js
38.7 fps /dist/mapbox-gl.js
```

**Comment by lucaswoj** [comment]
I'm seeing the same perf regression before any of the style refactoring hit `master` (97d91501e366b26493e0dc9b08c05683e3672eab). Please continue to review this PR while I look for the source of the problem.

**Comment by lucaswoj** [comment]
Now I'm getting normal results :+1: There must've been something stuck in my cache. (cache was cleared before running this test) 

```
duration: 3000
48.1 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.12.4/mapbox-gl.js
49.7 fps /dist/mapbox-gl.js
48.4 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.12.4/mapbox-gl.js
49.2 fps /dist/mapbox-gl.js
```

**Comment by mourner** [comment]
Looking great to me. But let's wait for @jfirebaugh @ansis to chime in.

File: js/data/symbol_bucket.js [hunk 5]
```diff
@@ -326,12 +338,12 @@ SymbolBucket.prototype.placeFeatures = function(collisionTile, buffers, collisio
...
 
-    var layout = this.layoutProperties;
+    var layout = this.layer.layout;
     var maxScale = collisionTile.maxScale;
 
-    elementGroups.glyph['text-size'] = layout['text-size'];
+    elementGroups.glyph['text-size'] = this.adjustedTextSize;
     elementGroups.glyph['text-font'] = layout['text-font'];
-    elementGroups.icon['icon-size'] = layout['icon-size'];
+    elementGroups.icon['icon-size'] = this.adjustedIconSize;
```
Comment (ansis) [inline js/data/symbol_bucket.js]: I like how you renamed and separated adjustedIconSize from the real layout properties. Renaming it here too might be clearer: `elementGroups.icon.adjustedIconSize = this.adjustedIconSize` etc. They get used [here](https://github.com/mapbox/mapbox-gl-js/blob/70f0677d5ceeb1bd699d3b459410469d911f3483/js/render/draw_symbol.js#L149)

**Comment by ansis** [comment]
Looks good to me! just that one naming suggestion

**Comment by lucaswoj** [comment]
Great idea @ansis. I renamed the properties in 1527c5b and will merge when that gets the green light from CI.

**Pull request merged**